dango/lending/src/execute.rs

```diff
@@ -0,0 +1,177 @@
+use {
+    crate::{DEBTS, MARKETS},
+    anyhow::{ensure, Ok},
+    dango_types::{
+        account_factory::QueryAccountRequest,
+        bank,
+        config::ACCOUNT_FACTORY_KEY,
+        lending::{ExecuteMsg, InstantiateMsg, Market, MarketUpdates, NAMESPACE},
+    },
+    grug::{Addr, Coin, Coins, Denom, Inner, Message, MutableCtx, Part, Response},
+    std::{collections::BTreeMap, str::FromStr},
+};
+
+#[cfg_attr(not(feature = "library"), grug::export)]
+pub fn instantiate(ctx: MutableCtx, msg: InstantiateMsg) -> anyhow::Result<Response> {
+    for (denom, _updates) in msg.markets {
+        MARKETS.save(ctx.storage, &denom, &Market {})?;
+    }
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), grug::export)]
+pub fn execute(ctx: MutableCtx, msg: ExecuteMsg) -> anyhow::Result<Response> {
+    match msg {
+        ExecuteMsg::UpdateMarkets(updates) => update_markets(ctx, updates),
+        ExecuteMsg::Deposit {} => deposit(ctx),
+        ExecuteMsg::Withdraw {} => withdraw(ctx),
+        ExecuteMsg::Borrow(coins) => borrow(ctx, coins),
+    }
+}
+
+fn update_markets(
+    ctx: MutableCtx,
+    updates: BTreeMap<Denom, MarketUpdates>,
+) -> anyhow::Result<Response> {
+    // Ensure only chain owner can update markets denoms.
+    ensure!(
+        ctx.sender == ctx.querier.query_config()?.owner,
+        "Only the owner can whitelist denoms"
+    );
+
+    for (denom, _updates) in updates {
+        MARKETS.save(ctx.storage, &denom, &Market {})?;
+    }
+
+    Ok(Response::new())
+}
+
+/// Ensures that the sender's account is not a margin account.
+fn ensure_sender_account_is_not_margin(ctx: &MutableCtx) -> anyhow::Result<()> {
+    let account_factory: Addr = ctx.querier.query_app_config(ACCOUNT_FACTORY_KEY)?;
+    ensure!(
+        !ctx.querier
+            .query_wasm_smart(account_factory, QueryAccountRequest {
+                address: ctx.sender,
+            })?
+            .params
+            .is_margin(),
+        "Margin accounts can't deposit or withdraw"
+    );
+    Ok(())
+}
+
+/// Ensures that the sender's account is a margin account.
+fn ensure_sender_account_is_margin(ctx: &MutableCtx) -> anyhow::Result<()> {
+    let account_factory: Addr = ctx.querier.query_app_config(ACCOUNT_FACTORY_KEY)?;
+    ensure!(
+        ctx.querier
+            .query_wasm_smart(account_factory, QueryAccountRequest {
+                address: ctx.sender,
+            })?
+            .params
+            .is_margin(),
+        "Only margin accounts can borrow and repay"
+    );
+    Ok(())
+}
+
+pub fn deposit(ctx: MutableCtx) -> anyhow::Result<Response> {
+    // Ensure margin accounts can't deposit
+    ensure_sender_account_is_not_margin(&ctx)?;
```

Comment:
Now thinking about this again, maybe we don't need to prevent margin accounts from depositing.

Even if we do, user can simply deposit from a spot account and send the LP tokens to the margin account.

There will be probably also be strong demand for margin accounts to earn yields in the lending pool.

Reply:
Yeah agree. I don't really see a strong reason to prevent them from depositing.